Optimize UNNEST

yingsu00 opened the pull request:
Updates on 01/21/2019:
- Removed ArrayOfRowUnnestBuilder
- Addressed other comments from Masha and Andrii

Updates on 12/28/2019:
- Added appendNull() to Block interface in 252dc6072b
- Removed UnnestBuilderUtils which contained insertNull to all Blocks
- Refactored Unnester to Interface in b9c03f49fa
- Removed GenericUnnestBuilder class and directly use UnnestBuilder
- Resolved other comments from @mbasmanova and @arhimondr 

Resolves https://github.com/prestodb/presto/issues/13751

Improve performance of UnnestOperator by using vectorized execution and bulk copy of the blocks. JMH benchmark shows 2.3x-10.x improvement over baseline (unit is us/op):

No nulls, no ordinality

Replicated     | Unnest                |         before        |   after | gain
-------------|-----------------|-----------------|-------|-----
bigint  | array(integer)         |                   1,477.6      |      167.0    | 8.8
bigint  |  array(varchar)        |                    1,594.9       |     165.8   |  9.6
bigint   | map(varchar,varchar)      |                1,535.4     |       187.5   |  8.2
bigint   | array(array(varchar))    |                 1,499.7     |       171.2   |  8.8
bigint   | array(row(varchar,varchar,varchar))    |   1,682.7          |  201.7    | 8.3
varchar  | array(integer)                |            1,457.0        |    172.8  |   8.4
varchar |  array(varchar)               |             1,483.1        |    170.7   |  8.7
varchar  | map(varchar,varchar)           |           1,586.4     |       203.9  |   7.8
varchar  | array(array(varchar))     |                1,489.8  |          169.0 |    8.8
varchar  | array(row(varchar,varchar,varchar))    |   1,681.0         |   204.8 |    8.2

No nulls, with ordinality

Replicated  | Unnest          |                       before      |       after   | gain
-------------|-----------------|-----------------|-------|-----
bigint  |  array(integer)        |                    1,590.6        |    284.4  |   5.6
bigint |   array(varchar)    |                        2,987.8  |          276.5 |    10.8
bigint   | map(varchar,varchar)  |                    1,685.4  |          303.2  |   5.6
bigint  |  array(array(varchar))    |                 1,613.3      |      310.9  |   5.2
bigint   | array(row(varchar,varchar,varchar))  |     1,865.8  |          343.8 |    5.4
varchar  | array(integer) |                           1,609.8 |           280.7 |    5.7
varchar |  array(varchar)    |                        1,605.9   |         291.9   |  5.5
varchar  | map(varchar,varchar)  |                    1,639.1 |           317.8 |    5.2
varchar  | array(array(varchar))   |                  1,593.8  |          315.1  |   5.1
varchar  | array(row(varchar,varchar,varchar)) |      1,840.4 |           341.5 |    5.4

No ordinality, 2 unnested columns

nullsRatio  | Replicated   |  Unnest       |               before    |       after |   gain
-------------|-----------------|-----------------|-------|-----|--------
0      |   bigint  |  array(bigint)+array(bigint)   |     2,153.7     |      581.5  |   3.7
0      |   bigint  |  array(varchar)+array(varchar)   |   4,799.6   |        615.4  |   7.8
0     |    varchar  | array(bigint)+array(bigint)   |     2,105.1     |      669.8   |  3.1
0      |   varchar |  array(varchar)+array(varchar)  |    4,839.5    |       620.5   |  7.8
0.2    |   bigint  |  array(bigint)+array(bigint)   |     1,884.3       |    523.3    | 3.6
0.2    |   bigint    | array(varchar)+array(varchar)  |    1,867.9       |    496.9    | 3.8
0.2     |  varchar  |  array(bigint)+array(bigint)    |    1,871.2         |  529.8    | 3.5
0.2     |  varchar   | array(varchar)+array(varchar)   |   1,884.2      |     516.2   |  3.7

Array of rows with 5% nulls in RowBlock

nullsRatio  | Replicated   |  Unnest            |                 before      |      after  |  gain
-------------|-----------------|-----------------|-------|-----|--------
0      |   bigint |   array(row(varchar,varchar,varchar))   |     6,462.5     |      777.5   |  8.3
0     |    varchar |  array(row(varchar,varchar,varchar))   |     6,720.0     |      702.7  |   9.6
0.2    |   bigint   | array(row(varchar,varchar,varchar))    |    1,732.1        |   720.1    | 2.4
0.2    |   varchar  | array(row(varchar,varchar,varchar))   |     1,747.5       |    743.8   |  2.3

```
== RELEASE NOTES ==

General Changes
* Improve performance of UnnestOperator

```

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java:
```diff
@@ -141,11 +141,13 @@ public static Block createStringsBlock(Iterable<String> values)
         return createStringsBlock(
                 IntStream.range(0, positionCount)
-                        .mapToObj(i -> allowNulls && i % 7 == 1 ? null : generateRandomStringWithLength(maxStringLength))
+                        .mapToObj(i -> nullRate > 0 && ThreadLocalRandom.current().nextDouble(1) < nullRate ? null : generateRandomStringWithLength(maxStringLength))
```
nit: consider extracting this logic into a helper class to avoid copy-paste

```
        ValuesWithNullsGenerator<String> generator = new ValuesWithNullsGenerator(nullRate, () -> generateRandomStringWithLength(maxStringLength));

        return createStringsBlock(
                IntStream.range(0, positionCount)
                        .mapToObj(i -> generator.next())
                        .collect(Collectors.toList()));

    private static final class ValuesWithNullsGenerator<T>
    {
        private final float nullRate;
        private final Supplier<T> supplier;

        private ValuesWithNullsGenerator(float nullRate, Supplier<T> supplier)
        {
            checkArgument (nullRate < 0 || nullRate > 1, "nullRate must be between 0 and 1 inclusive");

            this.nullRate = nullRate;
            this.supplier = requireNonNull(supplier, "supplier is null");
        }

        public T next()
        {
            return nullRate > 0 && ThreadLocalRandom.current().nextDouble(1) < nullRate ? null : supplier.get());
        }
    }
```

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/block/BlockAssertions.java:
```diff
@@ -821,4 +848,11 @@ private static String generateRandomStringWithLength(int length)
         return new String(array, UTF_8);
     }
+
+    private static void verifyNullRate(float nullRate)
```
this method can be replaced with `checkArgument (nullRate < 0 || nullRate > 1, "nullRate must be between 0 and 1 inclusive");` in the helper class

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -14,45 +14,37 @@
 {
-    private final Type type;
-    private Block source;
+    public Type type;
```
make `type` private final

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -14,45 +14,37 @@
-    private final Type type;
-    private Block source;
+    public Type type;
+    public Block source;
```
make `source` private and add protected getter

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -52,7 +60,14 @@
         private boolean closed;
 
-        public UnnestOperatorFactory(int operatorId, PlanNodeId planNodeId, List<Integer> replicateChannels, List<Type> replicateTypes, List<Integer> unnestChannels, List<Type> unnestTypes, boolean withOrdinality)
+        public UnnestOperatorFactory(
```
nit: would you move this refactoring into a separate commit?

yingsu00 replied:
Sure. Moved this refactor and method reordering into a separate commit "Refactor UnnestOperator"

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
     }
 
-    private int processCurrentPosition()
+    private static Unnester createUnnester(Type nestedType, boolean isLegacyUnnest)
```
looks like this method just moved? would you put mechanical refactorings/moves into a separate commit to make it easier to see material changes?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -14,32 +14,32 @@
-    private int currentPosition;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(Unnester.class).instanceSize();
+
+    protected UnnestBlockBuilder[] unnestBlockBuilders;
```
make unnestBlockBuilders final

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -14,32 +14,32 @@
+    protected int currentPosition;
+    protected int[] entryCounts = new int[0];
 
     protected Unnester(Type... types)
```
this constructor doesn't use the types; it just needs the total count; hence, let's change it to

```
    protected Unnester(int fieldCount)
    {
        unnestBlockBuilders = new UnnestBlockBuilder[fieldCount];
        currentPosition = 0;
    }
```

Then, change the subclasses to avoid re-allocating unnestBlockBuilders and re-initializing currentPosition:

```
    public ArrayOfRowsUnnester(Type... types)
    {
        super(types.length);
        this.fieldCount = types.length;
        for (int i = 0; i < fieldCount; i++) {
            unnestBlockBuilders[i] = new ArrayOfRowUnnestBlockBuilder(types[i], entryCounts);
        }
    }

    public MapUnnester(Type keyType, Type valueType)
    {
        super(2);
        unnestBlockBuilders[0] = new GenericUnnestBlockBuilder(keyType, entryCounts);
        unnestBlockBuilders[1] = new GenericUnnestBlockBuilder(valueType, entryCounts);
    }

    public ArrayUnnester(Type elementType)
    {
        super(1);
        unnestBlockBuilders[0] = new GenericUnnestBlockBuilder(elementType, entryCounts);
    }
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -14,32 +14,32 @@
+
+    protected UnnestBlockBuilder[] unnestBlockBuilders;
+    protected int currentPosition;
+    protected int[] entryCounts = new int[0];
```
term `count` here is inconsistent with pre-existing `getElementsLength`; consider renaming to `lengths`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -55,62 +55,51 @@ public final void resetInput(Block block)
         }
+
+        for (int i = 0; i < positionCount; i++) {
+            int currentEntryCount = getElementsLength(i);
```
currentEntryCount -> length

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
+        for (Unnester unnester : unnesters) {
+            unnester.populateEntryCounts(positionCount);
+            for (int i = 0; i < positionCount; i++) {
+                int curLen = unnester.getEntryCounts()[i];
```
don't abbreviate; perhaps, change Unnested.getLengths() method to `getLength(int position)` and write this like so:

```
maxEntries[i] = max(maxEntries[i], unnester.getLength(i));

    public int getLength(int position)
    {
        return lengths[position];
    }
```

yingsu00 replied:
getEntryLengths() is still required, therefore, I added the function getEntryLength(int position) as follows:

```
public int[] getEntryLengths()
    {
        return entryLengths;
    }
    
    public int getEntryLength(int position)
    {
        assert position >= 0 && position < entryLengths.length;
        return entryLengths[position];
    }
```

yingsu00 commented:
@mbasmanova Thank you for reviewing the PR and bringing up many good suggestions! I have addressed the comments and updated the PR.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
-    private static final int DEFAULT_CAPACITY = 64;
-    // See java.util.ArrayList for an explanation
-    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    static Block insertNull(Block block, Type type)
```
- What does this method do? My reading of the implementation is that it adds one null position at the end of the block. Perhaps, rename it to `appendNull`. 
- I'd extract this method into a separate commit. This is a self-contained unit-testable addition.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
-        // verify new size is within reasonable bounds
-        if (newSize < DEFAULT_CAPACITY) {
-            newSize = DEFAULT_CAPACITY;
+        int[] ids = new int[positionCount + 1];
```
perhaps, use `getIds`: `int[] ids = Arrays.copyOf(block.getRawIds(), positionCount + 1);`

yingsu00 replied:
Hi Masha, getRawIds() is package private and can't be accessed by the operator. After we add appendNull() to Block interface these won't be needed.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
-            if (newSize == currentSize) {
-                throw new IllegalArgumentException(format("Can not grow array beyond '%s'", MAX_ARRAY_SIZE));
+
+        NullElementFinder nullElementFinder = new NullElementFinder();
```
Using `NullElementFinder` is not beneficial here. Let's just write out the logic:

```
        int nullIndex = findNull(dictionary);

    private static int findNull(Block block)
    {
        int nullIndex = -1;
        if (block.mayHaveNull()) {
            for (int i = 0; i < block.getPositionCount(); i++) {
                if (block.isNull(i)) {
                    nullIndex = i;
                }
            }
        }
        return nullIndex;
    }
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
+        assert value.getPositionCount() == 1;
+
+        int[] ids = new int[positionCount + 1];
+        for (int i = 0; i < positionCount; i++) {
```
this loop is redundant; 0 is the default value for `int`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
+            ids[i] = 0;
+        }
+
+        NullElementFinder nullElementFinder = new NullElementFinder();
```
no need for NullElementFilter here

```
        int[] ids = new int[positionCount + 1];

        if (value.isNull(0)) {
            ids[positionCount] = 0;
        }
        else {
            value = appendNull(value, type);
            ids[positionCount] = 1;
        }
        return new DictionaryBlock(value, ids);
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
+        return new DictionaryBlock(value, ids);
+    }
+
+    private static Block insertNullToByteArrayBlock(ByteArrayBlock block)
```
Why not add `appendNull` method to the Block interface?

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/NullElementFinder.java:
```diff
@@ -0,0 +1,69 @@
+ * The result is cached with the first invocation of {@link #getNullElementIndex} after the reset. The
+ * cache can be invalidated by invoking {@link #resetCheck}.
+ */
+class NullElementFinder
```
This class is very leaky. It seem to be used only in UnnestBlockBuilder. I suggest to move this logic there.

mbasmanova commented:
@yingsu00 

> I have addressed the comments and updated the PR.

Somehow I still see `entryCount` throughout the PR. I think using `lengths` would be more consistent with the rest of the code base. (I don't think we need `entryLengths` because it is clear that `lengths` is about lengths of array or map elements.)

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -122,4 +115,20 @@ protected final UnnestBlockBuilder getBlockBuilder(int index)
+        return unnestBlockBuilders[index];
+    }
+
+    protected void setUnnestBlockBuilder(UnnestBlockBuilder unnestBlockBuilder, int index)
```
This method is used only in the constructors. It can be replaced with a constructor argument.

1. change constructor for Unnester to take a list of UnnestBlockBuilders
```
    protected Unnester(UnnestBlockBuilder... unnestBlockBuilders)
    {
        this.unnestBlockBuilders = requireNonNull(unnestBlockBuilders, "unnestBlockBuilders is null");
    }
```

2. Change constructor for ArrayUnnester

```
super(new GenericUnnestBlockBuilder(elementType));
```

3. Change constructor for MapUnnester

```
super(new GenericUnnestBlockBuilder(keyType), new GenericUnnestBlockBuilder(valueType));
```

4. Change constructor for ArrayOfRowsUnnester

```
    public ArrayOfRowsUnnester(Type... types)
    {
        super(createUnnestBlockBuilder(types));
        this.fieldCount = types.length;
    }

    private static UnnestBlockBuilder[] createUnnestBlockBuilder(Type[] types)
    {
        UnnestBlockBuilder[] builders = new UnnestBlockBuilder[types.length];

        for (int i = 0; i < types.length; i++) {
            builders[i] = new ArrayOfRowUnnestBlockBuilder(types[i]);
        }

        return builders;
    }
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayOfRowsUnnester.java:
```diff
@@ -117,4 +80,27 @@ protected int getElementsLength(int index)
         return columnarArray.getLength(index);
     }
+
+    protected boolean needToInsertNulls(int offset, int length, int maxTotalEntriesForBatch, int unnesterCount)
```
@Override

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -65,22 +58,22 @@ protected Block getElementsBlock(int channel)
 
-    @Override
-    protected int getElementsLength(int index)
+    protected boolean needToInsertNulls(int offset, int length, int maxTotalEntriesForBatch, int unnesterCount)
```
@Override

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/MapUnnester.java:
```diff
@@ -92,4 +66,18 @@ protected int getElementsLength(int index)
         return columnarMap.getEntryCount(index);
     }
+
+    protected boolean needToInsertNulls(int offset, int length, int maxTotalEntriesForBatch, int unnesterCount)
```
@Override

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -65,22 +58,22 @@ protected Block getElementsBlock(int channel)
+        }
+
+        int totalEntriesForBatch = columnarArray.getOffset(offset + length) - columnarArray.getOffset(offset);
+        if (totalEntriesForBatch < maxTotalEntriesForBatch) {
```
- totalEntriesForBatch -> totalEntries for readability
- replace `if` with `return`
- ditto in MapUnnester

```
        int totalEntries = columnarArray.getOffset(offset + length) - columnarArray.getOffset(offset);
        return totalEntries < maxTotalEntriesForBatch;
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayOfRowsUnnester.java:
```diff
@@ -37,19 +34,20 @@
-    private int nullRowsEncountered;
+    private ColumnarArray columnarArray;
+    private ColumnarRow columnarRow;
+    private Block nullCheckBlock;
```
nullCheckBlock variable is not needed

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayOfRowsUnnester.java:
```diff
@@ -117,4 +80,27 @@ protected int getElementsLength(int index)
+
+    protected boolean needToInsertNulls(int offset, int length, int maxTotalEntriesForBatch, int unnesterCount)
+    {
+        int rowStartPosition = columnarArray.getOffset(offset);
```
perhaps, simplify a bit:

```
        int start = columnarArray.getOffset(offset);
        int end = columnarArray.getOffset(offset + length);
        int totalEntries = end - start;

        if (totalEntries < maxTotalEntriesForBatch) {
            return true;
        }

        if (columnarRow.getNullCheckBlock().mayHaveNull()) {
            for (int i = start; i < end; i++) {
                if (columnarRow.isNull(i)) {
                    return true;
                }
            }
        }

        return false;
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -52,65 +51,59 @@ public final void resetInput(Block block)
     }
 
-    public final int getCurrentUnnestedLength()
+    public abstract int getChannelCount();
```
- getChannelCount() method is not needed; the result can be computed as unnestBlockBuilders.length

```
    private final int channelCount;

    protected Unnester(UnnestBlockBuilder... unnestBlockBuilders)
    {
        this.unnestBlockBuilders = requireNonNull(unnestBlockBuilders, "unnestBlockBuilders is null");
        this.channelCount = unnestBlockBuilders.length;
    }
```

yingsu00 replied:
It's still needed by the operator. But I have changed the implementation of it to
```
public int getChannelCount()
    {
        return channelCount;
    }
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -62,204 +53,75 @@ public UnnestBlockBuilder(Type type)
-        clearCurrentOutput();
 
-        return outputBlock;
+        return new DictionaryBlock(maxTotalEntriesForBatch, source, ids);
```
Why is DictionaryBlock useful here? Why not just return a view into source block?

```
        Block output = source.getRegion(sourcePosition, batchSize);
        sourcePosition += batchSize;
        return output;
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -62,204 +53,75 @@ public UnnestBlockBuilder(Type type)
-     * - {@code usingCopiedBlock} is false
-     */
-    private void appendId(int sourcePosition)
+    protected NullElementFinder getNullFinder()
```
This method is not needed. Just add `nullIndex` parameter to `buildOutputBlockWithNullsByDictionary`.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayOfRowUnnestBlockBuilder.java:
```diff
@@ -0,0 +1,77 @@
+import static com.google.common.base.Verify.verify;
+import static java.util.Objects.requireNonNull;
+
+class ArrayOfRowUnnestBlockBuilder
```
It seems to me that you don't need this class at all. Just add "optional" nullCheckBlock to UnnestBlockBuilder

```
    public Block buildOutputBlockWithNulls(int[] maxEntries, int offset, int positionCount, int batchSize)
    {
        int nullIndex = nullFinder.getNullElementIndex();
        if (nullIndex == INVALID_NULL_POSITION) {
            source = insertNull(source, type);
            nullIndex = source.getPositionCount() - 1;
            nullFinder.setNullElementIndex(nullIndex);
        }

        int position = 0;
        int[] ids = new int[batchSize];
        for (int i = 0; i < positionCount; i++) {
            int length = lengths[offset + i];
            for (int j = 0; j < length; j++) {
                if (nullCheckBlock != null && nullCheckBlock.isNull(nullCheckBlockPosition++)) {
                    ids[position++] = nullIndex;
                }
                else {
                    ids[position++] = sourcePosition++;
                }
            }

            int maxLength = maxEntries[offset + i];
            for (int j = length; j < maxLength; j++) {
                ids[position++] = nullIndex;
            }
        }

        return new DictionaryBlock(batchSize, source, ids);
    }
```

yingsu00 replied:
The reason behind it was trying to avoid if checks in the loop for GenericUnnestBlockBuilder so the loop could be optimized. Mixing the two loops would regress the GenericUnnestBlockBuilder by over 20%.

mbasmanova replied:
@yingsu00 Ok. Let's not mix the loops. Just move the if check outside of the loop. I still don't think we need a separate class and would like to avoid getSourcePosition/setSourcePosition pattern.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -62,204 +53,75 @@ public UnnestBlockBuilder(Type type)
-     * Get number of positions in the output block being prepared
-     */
-    public int getPositionCount()
+    protected int getSourcePosition()
```
All of these getter won't be needed once you add `nullCheckBlock`.

yingsu00 replied:
As I explained above, the existence of ArrayOfRowUnnestBlockBuilder was for performance purpose. I'd like to have the two implementations of buildOutputBlockWithNullsByDictionary separate.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/OrdinalityBlockBuilder.java:
```diff
@@ -0,0 +1,42 @@
+    private OrdinalityBlockBuilder()
+    {}
+
+    public static Block buildOutputBlock(int[] maxEntries, int offset, int length, int totalEntriesForBatch)
```
I'd inline this method and remove this class.

yingsu00 replied:
Moved the method to UnnestOperator.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ReplicatedBlockBuilder.java:
```diff
@@ -30,48 +26,25 @@
     {
-        this.source = requireNonNull(block, "block is null");
+        source = requireNonNull(block, "block is null");
+        sourcePosition = 0;
```
this assignment is redundant; 0 is the default value for int

yingsu00 replied:
@mbasmanova This is necessary because we need to reset sourcePosition to 0 every time a new page comes in.

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
+
+        for (int i = 0; i < unnestTypes.size(); i++) {
+            int inputChannel = unnestChannels.get(i);
+            Block unnestChannelInputBlock = currentPage.getBlock(inputChannel);
```
unnestChannelInputBlock  -> newInputBlock

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
+        int positionCount = currentPage.getPositionCount();
+        maxEntries = ensureCapacity(maxEntries, positionCount, SMALL, INITIALIZE);
+
+        for (Unnester unnester : unnesters) {
```
with `lengths` this block reads better:

```
        for (Unnester unnester : unnesters) {
            unnester.populateLengths(positionCount);
            for (int i = 0; i < positionCount; i++) {
                maxEntries[i] = max(maxEntries[i], unnester.getLength(i));
            }
        }
```

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
-    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    static Block insertNull(Block block, Type type)
+    {
+        if (block instanceof DictionaryBlock) {
```
Actually I'm not sure why this class is needed.

It doesn't provide any bulk-copy functionality. Why not simply create a new block builder by calling the `Type#createBlockBuilder` and copy the block into a new block builder?

Also we can have a `Type#createBlockBuilder(Block ...)` that creates a block builder based on the copy of block (and does that copy more efficient if it is proven that reconstructing the block using a BlockBuilder negatively impacts the performancce)

yingsu00 replied:
@arhimondr  Copying the block using these tight loops is 5-6x faster than calling `type.appendTo(block, j, blockBuilder)` or `blockBuilder.writeXXX(block.getXXX(i))` therefore I call it bulk copy. 

Let me use an example to see if I understand you correctly: You suggested adding `Type#createBlockBuilder(..., Block block)`, and add a new constructor to each BlockBuilder. E.g.
AbstractLongType:
```
@Override
    public final BlockBuilder createBlockBuilder(Block block, BlockBuilderStatus blockBuilderStatus)
    {
        int maxBlockSizeInBytes;
        if (blockBuilderStatus == null) {
            maxBlockSizeInBytes = PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
        }
        else {
            maxBlockSizeInBytes = blockBuilderStatus.getMaxPageSizeInBytes();
        }
        return new LongArrayBlockBuilder(
                blockBuilderStatus,
                block);
    }
```
Then add a new constructor of `LongArrayBlockBuilder`
```
    public LongArrayBlockBuilder(@Nullable BlockBuilderStatus blockBuilderStatus, Block block)
    {
        this.blockBuilderStatus = blockBuilderStatus;

        int positionCount = block.getPositionCount();
        values = new long[positionCount + 1];
        for (int i = 0; i < positionCount; i++) {
            values[i] = block.getLong(i);
        }

        valueIsNull = copyIsNulls(block);

        updateDataSize();
    }
```
Is this correct?

This should work well, but it still requires copying. I'd prefer adding a appendNull() to Block, so that in many places we could avoid copying at all. It was my first choice but I was afraid it would be more controversial than doing memory copy. But since @mbasmanova also suggested it I'd do it that way. Are you ok with it?

arhimondr commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -62,204 +53,75 @@ public UnnestBlockBuilder(Type type)
-        }
-        else {
-            appendId(index);
+        if (nullFinder.getNullElementIndex() == INVALID_NULL_POSITION) {
```
`INVALID_NULL_POSITION` -> `NULL_NOT_FOUND`?

arhimondr commented on presto-main/src/test/java/com/facebook/presto/operator/unnest/TestUnnesterUtil.java:
```diff
@@ -13,236 +13,297 @@
 import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
 
 public class TestUnnesterUtil
```
`UnnesterTestingUtils`

yingsu00 commented:
> @yingsu00
> 
> > I have addressed the comments and updated the PR.
> 
> Somehow I still see `entryCount` throughout the PR. I think using `lengths` would be more consistent with the rest of the code base. (I don't think we need `entryLengths` because it is clear that `lengths` is about lengths of array or map elements.)

Renamed them to lengths. However it may be confusing with the "int length" in the following code:

```
protected Block buildOutputBlockWithNullsByDictionary(int[] maxEntries, int offset, int length, int maxTotalEntriesForBatch)
    {
        int nullIndex = nullFinder.getNullElementIndex();
        verify(nullIndex != INVALID_NULL_POSITION, "nullIndex is -1");

        int position = 0;
        int[] ids = new int[maxTotalEntriesForBatch];
        for (int i = 0; i < length; i++) {
            int curEntries = lengths[offset + i];
            for (int j = 0; j < curEntries; j++) {
                ids[position++] = sourcePosition++;
            }

            int curMaxEntries = maxEntries[offset + i];
            for (int j = curEntries; j < curMaxEntries; j++) {
                ids[position++] = nullIndex;
            }
        }

        return new DictionaryBlock(maxTotalEntriesForBatch, source, ids);
    }
```

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
     }
 
-    private void prepareForNewOutput(PageBuilderStatus pageBuilderStatus)
+    private void calculateMaxEntries()
```
@arhimondr Thanks for the suggestion. As suggested by @mbasmanova, we renamed `entryCount` to `length`. `ExpansionFactor` is good too but might be too long. I'll rename it to `calculateMaxLengths` to make them consistent.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
+        int totalEntries = 0;
+        for (int i = currentPosition; i < positionCount; i++) {
+            int currentEntry = maxEntries[i];
+            if (totalEntries + currentEntry >= MAX_ROWS_PER_BLOCK) {
```
Thanks @arhimondr. Original way is shorter, but to make it easier to understand I updated it as you suggested:
```
private int calculateNextBatchSize()
    {
        int positionCount = currentPage.getPositionCount();

        int totalLengths = 0;
        int position = currentPosition;
        while (position < positionCount) {
            int length = maxLengths[position];
            if (totalLengths + length >= MAX_ROWS_PER_BLOCK) {
                break;
            }
            totalLengths += length;
            position++;
        }

        // grab at least a single position
        if (position == currentPosition) {
            currentBatchTotalLength = maxLengths[currentPosition];
            return 1;
        }

        currentBatchTotalLength = totalLengths;
        return position - currentPosition;
    }
```

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
 
         if (withOrdinality) {
-            outputBlocks[offset] = ordinalityBlockBuilder.build();
+            outputBlocks[channel] = buildOutputBlock(maxEntries, currentPosition, batchSize, currentBatchTotalEntries);
```
As both of you suggested, I moved the method to `UnnestOperator` and renamed it to `buildOrdinalityOutputBlock`, so now it's like this:
```
if (withOrdinality) {
            outputBlocks[channel] = buildOrdinalityOutputBlock(maxLengths, currentPosition, batchSize, currentBatchTotalLength);
        }
```

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperator.java:
```diff
@@ -196,111 +199,118 @@ public Page getOutput()
-        else {
-            throw new IllegalArgumentException("Cannot unnest type: " + nestedType);
+        long size = INSTANCE_SIZE + sizeOf(maxEntries);
+        for (Unnester unnester : unnesters) {
```
@arhimondr ReplicatedBlockBuilder doesn't hold any internal buffers. If we want to account for it, then it's just the object size which is very small number. It holds reference to incoming blocks, but I don't see incoming blocks are accounted in other operators either. Is there any example incoming pages are accounted by operators?

yingsu00 replied:
I added the currentPage to UnnestOperator.getRetainedSizeInBytes():
```
private long getRetainedSizeInBytes()
    {
        long size = INSTANCE_SIZE + sizeOf(maxLengths) + currentPage.getRetainedSizeInBytes();
        for (Unnester unnester : unnesters) {
            size += unnester.getRetainedSizeInBytes();
        }
        return size;
    }
```

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/Unnester.java:
```diff
@@ -122,4 +115,20 @@ protected final UnnestBlockBuilder getBlockBuilder(int index)
      */
     protected abstract int getElementsLength(int index);
+
+    protected abstract boolean needToInsertNulls(int offset, int length, int maxTotalEntriesForBatch, int unnesterCount);
```
It was intended to be an optimization but the difference is not visible. So I removed it as you suggested.

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java:
```diff
@@ -13,32 +13,304 @@
+        boolean[] valueIsNull = new boolean[positionCount + 1];
+        if (block.mayHaveNull()) {
+            for (int i = 0; i < positionCount; i++) {
+                valueIsNull[i] = block.isNull(i);
```
@arhimondr  Yes but the isNull array in Blocks is not exposed and that's why we had to copy it this way. I'm going to add `appendNull` to `Block`, which avoids copy and avoid breaking connector encapsulation. Please let me know if this is ok.

yingsu00 commented:
@mbasmanova and @arhimondr Happy new year! Thank you very much for your careful review! I have addressed your comments. The major updates in this version include:

- Added appendNull() to Block interface in 252dc60
- Removed UnnestBuilderUtils which contained insertNull to all Blocks
- Refactored Unnester to Interface in b9c03f4
- Removed GenericUnnestBuilder class and directly use UnnestBuilder

I still kept ArrayOfRowUnnestBlockBuilder because of performance reasons. Mixing the two loops regress Array and Map for 20%. 

Will you be able to review the PR again? Appreciate it!

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/block/Block.java:
```diff
@@ -305,4 +305,9 @@ default Block getLoadedBlock()
     }
+
+    /**
+     * Returns a block that has an appended null at the end. The original block won't be modified.
```
@yingsu00 Would you clarify what happens if original block already has a null?

yingsu00 replied:
Modified it to
```
   /**
     * Returns a block that has an appended null at the end, no matter if the original block has null or not. 
     * The original block won't be modified.
     */
```

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/block/BlockUtil.java:
```diff
@@ -42,6 +43,17 @@ static void checkArrayRange(int[] array, int offset, int length)
 
+    static void checkArrayRange(boolean[] array, int offset, int length)
+    {
+        if (array == null) {
```
the other checkArrayRange method on L38 doesn't allow null input; is there any particular reason to allow nulls here? Would it be possible to move null check to the caller and make this method throw on null input like the one one?

yingsu00 replied:
The blocks may not always have the null array. When it doesn't have null, there is no need to have the isNull array in the block. I modified it to the following:
```
static void checkArrayRange(boolean[] array, int offset, int length)
    {
        requireNonNull(array, "array is null");
        if (offset < 0 || length < 0 || offset + length > array.length) {
            throw new IndexOutOfBoundsException(format("Invalid offset %s and length %s in array with %s elements", offset, length, array.length));
        }
    }

static boolean[] appendNullToIsNullArray(@Nullable boolean[] isNull, int offsetBase, int positionCount)
    {
        int desiredLength = offsetBase + positionCount + 1;
        boolean[] newIsNull = new boolean[desiredLength];
        if (isNull != null) {
            checkArrayRange(isNull, offsetBase, positionCount);
            System.arraycopy(isNull, 0, newIsNull, 0, desiredLength - 1);
        }
        newIsNull[desiredLength - 1] = true;
        return newIsNull;
    }
```

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/block/BlockUtil.java:
```diff
@@ -216,4 +228,88 @@ public static boolean internalPositionInRange(int internalPosition, int offset,
+        return buffer;
+    }
+
+    static int findNull(Block block)
```
this method and NULL_NOT_FOUND variable are used only in DictionaryBlock; move them there and inline

mbasmanova commented on presto-main/src/test/java/com/facebook/presto/block/AbstractTestBlock.java:
```diff
@@ -199,7 +203,12 @@ else if (type == MethodHandle.class) {
     private <T> void assertBlockPositions(Block block, Supplier<BlockBuilder> newBlockBuilder, T[] expectedValues)
     {
-        assertEquals(block.getPositionCount(), expectedValues.length);
+        try {
```
I don't understand this change. Would you explain?

yingsu00 replied:
@mbasmanova Oh I'm sorry, this was for debugging purpose. I've removed the change.

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java:
```diff
@@ -540,4 +542,23 @@ public boolean isNullUnchecked(int internalPosition)
+    public Block appendNull()
+    {
+        int desiredLength = idsOffset + positionCount + 1;
+        int[] newIds = Arrays.copyOf(ids, desiredLength);
```
elsewhere you are using ensureCapacity helper method; why not here?

yingsu00 replied:
@mbasmanova This is because I didn't want to modify the original block. For DictionaryBlock, we always need to re-allocate ids array since we need to add one more element to it. If we modify the original ids array, then other block views that rely on this block get modified, which may cause problems. `ensureCapacity()` does not allocate new array if the array size is big enough so I just used `Arrays.copyOf()`.

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/block/RunLengthEncodedBlock.java:
```diff
@@ -375,4 +375,21 @@ public boolean isNullUnchecked(int internalPosition)
+        int[] ids = new int[positionCount + 1];
+
+        Block dictionary = value;
+        if (value.isNull(0)) {
```
perhaps, make an RLE block here; no need to wrap in Dictionary

yingsu00 replied:
Made it to the following:
```
@Override
    public Block appendNull()
    {
        if (value.isNull(0)) {
            return new RunLengthEncodedBlock(value, positionCount + 1);
        }
        else {
            Block dictionary = value.appendNull();
            int[] ids = new int[positionCount + 1];
            ids[positionCount] = 1;
            return new DictionaryBlock(dictionary, ids);
        }
    }
```

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/block/BlockUtil.java:
```diff
@@ -216,4 +228,88 @@ public static boolean internalPositionInRange(int internalPosition, int offset,
+        return newOffsets;
+    }
+
+    public static byte[] ensureCapacity(byte[] buffer, int capacity)
```
Add a comment that explains that this method will preserve the original data.

yingsu00 replied:
Added the following comment:
```
/**
     * Returns a short array of size capacity if the input buffer is null or smaller than the capacity.
     * If the input buffer is not null, the original values in buffer will be preserved.
     */
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -46,33 +50,51 @@ public int getChannelCount()
     {
-        checkState(channel == 0, "index is not 0");
-        return columnarArray.getElementsBlock();
+        boolean nullRequired = needToInsertNulls(startPosition, batchSize, currentBatchTotalEntries);
```
let's inline needToInsertNulls:

```
boolean nullRequired = columnarArray.getOffset(startPosition + batchSize) - columnarArray.getOffset(startPosition) < currentBatchTotalEntries;
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -28,15 +29,18 @@
     public ArrayUnnester()
     {
-        super(new UnnestBlockBuilder());
+        this.unnestBlockBuilder = new UnnestBlockBuilder();
```
Combine with member variable declaration:

```
private final UnnestBlockBuilder unnestBlockBuilder = new UnnestBlockBuilder();
```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -46,33 +50,51 @@ public int getChannelCount()
 
     @Override
-    protected void resetColumnarStructure(Block block)
+    public int[] populateLengths(int positionCount)
```
I find this method awkward. Looking at how it is used, I'm seeing that it is always called after resetInput. Hence, I suggest to fold this logic into resetInput. I also see that UnnestBlockBuilder uses lengths array only in buildOutputBlockWithNulls method. Hence, I suggest to remove setLengths methods and pass lengths to buildOutputBlockWithNulls method directly.

The interface would looks like this:

```
void resetInput(Block block, int positionCount);

int[] getLengths();
```

and the implementation:

```
    @Override
    public void resetInput(Block block, int positionCount)
    {
        requireNonNull(block, "block is null");

        columnarArray = toColumnarArray(block);
        unnestBlockBuilder.resetInputBlock(columnarArray.getElementsBlock());

        lengths = ensureCapacity(lengths, positionCount);
        for (int i = 0; i < positionCount; i++) {
            lengths[i] = columnarArray.getLength(i);
        }
    }

    @Override
    public int[] getLengths()
    {
        return lengths;
    }

    @Override
    public Block[] buildOutputBlocks(int[] maxEntries, int startPosition, int batchSize, int currentBatchTotalEntries)
    {
        boolean nullRequired = columnarArray.getOffset(startPosition + batchSize) - columnarArray.getOffset(startPosition) < currentBatchTotalEntries;

        Block[] outputBlocks = new Block[1];
        if (nullRequired) {
            outputBlocks[0] = unnestBlockBuilder.buildOutputBlockWithNulls(maxEntries, startPosition, batchSize, currentBatchTotalEntries, lengths);
        }
        else {
            outputBlocks[0] = unnestBlockBuilder.buildOutputBlockWithoutNulls(currentBatchTotalEntries);
        }

        return outputBlocks;
    }
```

yingsu00 commented:
> Consider adding some unit tests at least for RLE and Dictionary blocks. See TestRunLengthEncodedBlock and TestDictionaryBlock.

Hi Masha, I already added test to AbstractTestBlock#assertBlock that covers all block types including RLE and Dictionary blocks. Did you mean more tests for these two type of blocks?

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -62,184 +44,89 @@ public UnnestBlockBuilder(Type type)
      * cache can be invalidated by invoking {@link #resetCheck}.
      */
-    private class NullElementFinder
+    static class NullElementFinder
```
@arhimondr In UnnestBlockBuilder it is using UnnestBlockBuilder.NullElementFinder.NULL_NOT_FOUND. If NullElementFinder is private NULL_NOT_FOUND cannot be reference outside of NullElementFinder.

yingsu00 commented:
Thank you @arhimondr and @mbasmanova ! I've been doing verifier runs and so far so good. While I'm doing more verifier runs, @mbasmanova do you still have more comments? I removed ArrayOfRowUnnestBuilder and addressed all of other comments. Thank you!

mbasmanova commented:
@yingsu00 Ying, I haven't looked at the code, but the list of commits doesn't make sense to me. I see "Optimize UnnestOperator" commit that uses Block.appendNull method come before "Add appendNull() to Block interface" commit. I also see a couple of refactoring commits coming later. A natural order of commits would be: (1) refactoring comes first; (2) supporting changes (e.g. new appendNull method) next; (3) the final optimization/improvement last. Would you re-arrange commits in that way?

yingsu00 commented:
@mbasmanova hi Masha, the "Add appendNull() to Block interface" was before the "Optimize UnnestOperator", please see the attached image
<img width="664" alt="Screen Shot 2020-01-31 at 12 55 43 AM" src="https://user-images.githubusercontent.com/33299678/73525631-742d5380-43c4-11ea-8908-6a3d0ce2b2a3.png">
The github webpage shows the order differently, but if you fetch the PR you will see the real order.

I checked the original code and found it is not very straightforward to put the "Refactor Unnester" before the optimization. In the optimization, I removed a lot of methods from these classes, which made the refactoring easier and make sense. In the new Unnester interface, there're only 5 methods left and the interface looked clean. The original Unnester was an abstract class that contains many methods and was not well designed. It doesn't make too much sense to me to refactor (make it interface) before the optimization.

Similarly, the "Refactor UnnestOperator" was to reorder the methods after deleting a lot of methods in the "Optimize UnnestOperator" commit. I think this is the easiest order for you to see the changes.

If I misunderstood your comment, could you please explain a bit more what exactly you'd like me to do? Thank you!

mbasmanova commented:
@yingsu00 

> The github webpage shows the order differently, but if you fetch the PR you will see the real order.

This is a known issue and here is a fix: https://gist.github.com/findepi/7abf17c8a26d3b74e8bb9527caadfdbe Please apply it so that GitHub shows the commits in the right order.

yingsu00 commented:
> @yingsu00
> 
> > The github webpage shows the order differently, but if you fetch the PR you will see the real order.
> 
> This is a known issue and here is a fix: https://gist.github.com/findepi/7abf17c8a26d3b74e8bb9527caadfdbe Please apply it so that GitHub shows the commits in the right order.

@mbasmanova Thank you Masha! It worked like a charm :)

mbasmanova commented:
@yingsu00 I'm seeing that "Optimize UnnestOperator" commit adds ArrayOfRowUnnestBlockBuilder.java file, but "Refactor Unnester" commit deletes it. It is very confusing to add and delete a file in the same PR. Is this really necessary?

yingsu00 commented:
@mbasmanova Masha, I reworked the commits as you requested, now the "Refactor UnnestOperator" and "Refactor Unnester" are before "Optimize Unnest". "Refactor UnnestOperator" reorders the methods/variables in UnnestOperator. "Refactor Unnest" changed the abstract class Unnest to an Interface and removed unnecessary methods. Hope this is easier for you to review. Thank you!

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -43,44 +49,50 @@ public int getChannelCount()
 
     @Override
-    protected Block getElementsBlock(int channel)
+    public final int getCurrentUnnestedLength()
```
drop `final`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -43,44 +49,50 @@ public int getChannelCount()
 
     @Override
-    protected void processCurrentPosition(int requiredOutputCount)
+    public final void processCurrentAndAdvance(int requiredOutputCount)
```
drop `final`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/ArrayUnnester.java:
```diff
@@ -43,44 +49,50 @@ public int getChannelCount()
 
     @Override
-    protected int getElementsLength(int index)
+    public final Block[] buildOutputBlocksAndFlush()
```
drop `final`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/MapUnnester.java:
```diff
@@ -27,69 +29,81 @@
-        return 2;
+        requireNonNull(block, "block is null");
+
+        this.columnarMap = toColumnarMap(block);
```
drop `this.`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/MapUnnester.java:
```diff
@@ -27,69 +29,81 @@
 
     @Override
-    protected void resetColumnarStructure(Block block)
+    public final int getCurrentUnnestedLength()
```
drop `final`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/MapUnnester.java:
```diff
@@ -27,69 +29,81 @@
 
     @Override
-    protected int getElementsLength(int index)
+    public final Block[] buildOutputBlocksAndFlush()
```
drop `final`

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestBlockBuilder.java:
```diff
@@ -14,232 +14,118 @@
             this.checkedForNull = false;
-            this.nullElementPosition = -1;
+            this.nullElementPosition = NULL_NOT_FOUND;
+            this.source = source;
```
nit: requireNonNull?

yingsu00 commented:
@mbasmanova @arhimondr Hi Masha and Andrii, after dozens of verifier runs which include over 30,000 unnest queries, we are finally cleared with the verifier runs in all regions. I think this PR is ready to merge. Thank you very much for your careful review and comments. Please let me know if there's more things need to be done.

caithagoras commented:
@yingsu00 @mbasmanova 
master is broken on Travis due to this PR.
```
com.facebook.presto.operator.unnest.TestUnnestOperator.testUnnestTwoArrayOfRowUnnesters
com.facebook.presto.operator.unnest.TestUnnestOperator.testUnnestSingleArrayOfRowUnnester
com.facebook.presto.operator.unnest.TestUnnestOperator.testUnnestMultipleUnnesters
```

yingsu00 commented:
> TwoArrayOfRowUnnesters

I'm looking

yingsu00 commented:
> > TwoArrayOfRowUnnesters
> 
> I'm looking

Fix is https://github.com/prestodb/presto/pull/14416
